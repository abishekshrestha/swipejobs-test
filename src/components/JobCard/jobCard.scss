.job-card {
  display: flex;
  flex-direction: column;
  background-color: #fff;
  border-radius: 6px;
  overflow: hidden;
  min-height: calc(100vh - 100px);
  p {
    margin: 0 0 4px 0;
  }

  &-container {
    padding: 10px 20px;

    &--content {
      display: flex;
      flex-direction: column;
      justify-content: space-between;
      flex-grow: 1;
    }
  }

  .image-style {
    width: 100%;
  }
  .job-description {
    &__title {
      font-weight: 800;
    }
    &__company {
      font-size: 14px;
      font-weight: 600;
    }
  }

  .distance-rate-info {
    display: flex;
    justify-content: space-between;
    background-color: #2fd5aa;

    &__title {
      font-size: 11px;
      font-weight: 700;
      padding-bottom: 4px;
    }

    .text-align-end {
      text-align: end;
    }

    &__value {
      color: white;
      font-size: 28px;
      font-weight: 700;
    }

    &__dollar {
      font-size: 14px;
      position: relative;
      top: -12px;
    }
  }

  .job-item {
    display: flex;
    padding: 16px 0;
    align-items: flex-start;
    border-bottom: 1px solid #ddd;

    &:last-child {
      border-bottom: none;
    }

    &__location {
      text-decoration: none;
      color: currentColor;
    }

    &__icon {
      padding-right: 16px;
      padding-top: 8px;
    }

    &__chevron {
      stroke: white;
      align-self: center;
    }

    &__title {
      font-size: 14px;
      font-weight: 700;
    }
    &__content {
      font-size: 14px;
      flex: auto;
      &--small {
        font-size: 11px;
      }
    }
    &__date {
      text-transform: uppercase;
    }
  }

  .button-section {
    display: flex;
    align-items: center;
    gap: 6px;
    padding-top: 10px;
    min-height: 53px;

    .button {
      padding: 16px 24px;
      border-radius: 4px;
      width: 100%;
      font-size: 16px;
      &-light {
        color: #bdbdbd;
        background-color: white;
        border: #bdbdbd solid 1px;
      }
      &-dark {
        color: white;
        background-color: black;
        border: black solid 1px;
      }
    }
  }
}
