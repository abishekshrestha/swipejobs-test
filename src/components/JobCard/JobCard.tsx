import AccountCircleIcon from '@mui/icons-material/AccountCircle';
import ArrowForwardIosIcon from '@mui/icons-material/ArrowForwardIos';
import CalendarMonthIcon from '@mui/icons-material/CalendarMonth';
import HandymanIcon from '@mui/icons-material/Handyman';
import LocationOnIcon from '@mui/icons-material/LocationOn';
import { formatToTimeZone } from 'date-fns-timezone';
import { IJobDetail } from '../../models/job';
import Spinner from '../Spinner/Spinner';
import './jobCard.scss';

interface IProps extends IJobDetail {
  handleAcceptButton?: () => void;
  handleRejectButton?: () => void;
  isLoading?: boolean;
}

const JobCard = ({
  jobTitle,
  company,
  milesToTravel,
  wagePerHourInCents,
  shifts,
  requirements,
  handleAcceptButton,
  handleRejectButton,
  isLoading,
}: IProps) => {
  const { address, name, reportTo } = company || {};

  /**
   * Calculate wage in dollar value
   */
  const wagePerHourInDollar = Number(wagePerHourInCents) / 100;

  /**
   * Date format
   */
  const dateFormatStart = 'MMM D, ddd h:mm a';
  const dateFormatEnd = 'h:mm a z';

  /**
   * Function to format the date depending upon timezone
   */
  const convertDateTimezone = (
    date: string,
    format: string = dateFormatStart
  ) =>
    formatToTimeZone(new Date(date), format, {
      timeZone: company?.address?.zoneId,
    });

  /**
   * Format phone number. E.g. (123) 456 7890
   */
  const formatPhoneNumber = (phoneNumberString: string) => {
    const cleaned = ('' + phoneNumberString).replace(/\D/g, '');
    const match = cleaned.match(/^(\d{3})(\d{3})(\d{4})$/);
    if (match) {
      return `(${match[1]}) ${match[2]} ${match[3]}`;
    }
    return null;
  };

  return (
    <div className='job-card'>
      <img className='image-style' src={jobTitle?.imageUrl} alt='job_image' />
      <div className='job-description job-card-container'>
        <p className='job-description__title'>{jobTitle?.name}</p>
        <p className='job-description__company'>{name}</p>
      </div>
      <div className='distance-rate-info job-card-container'>
        <div>
          <div className='distance-rate-info__title'>Distance</div>
          <div className='distance-rate-info__value'>
            {Number(milesToTravel).toFixed(1)} miles
          </div>
        </div>
        <div>
          <div className='distance-rate-info__title text-align-end'>
            Hourly Rate
          </div>
          <div className='distance-rate-info__value text-align-end'>
            <sup className='distance-rate-info__dollar'>$</sup>
            {wagePerHourInDollar.toFixed(2)}
          </div>
        </div>
      </div>
      <div className='job-card-container job-card-container--content'>
        <div>
          {shifts && (
            <div className='job-item'>
              <CalendarMonthIcon className='job-item__icon' />
              <div className='job-item__content'>
                <p className='job-item__title'>Shift Dates</p>
                {shifts?.map((item, idx) => (
                  <div className='job-item__date' key={idx}>
                    <div>
                      {convertDateTimezone(item?.startDate, dateFormatStart)} -{' '}
                      {convertDateTimezone(item?.endDate, dateFormatEnd)}
                    </div>
                  </div>
                ))}
              </div>
            </div>
          )}

          {address?.formattedAddress && (
            <a
              href='https://map.google.com'
              className='job-item job-item__location'
              target='_blank'
              rel='noopener noreferrer'
            >
              <LocationOnIcon className='job-item__icon' />
              <div className='job-item__content'>
                <p className='job-item__title'>Location</p>
                <p className=''>{address?.formattedAddress}</p>
                {milesToTravel && (
                  <p className='job-item__content--small'>
                    {Number(milesToTravel).toFixed(2)} miles from your job
                    search location
                  </p>
                )}
              </div>
              <ArrowForwardIosIcon className='job-item__chevron' />
            </a>
          )}

          {requirements && (
            <div className='job-item'>
              <HandymanIcon className='job-item__icon' />
              <div className='job-item__content'>
                <p className='job-item__title'>Requirements</p>
                {requirements?.map((item, idx) => (
                  <p key={idx}>{item}</p>
                ))}
              </div>
            </div>
          )}

          {reportTo && (
            <div className='job-item'>
              <AccountCircleIcon className='job-item__icon' />
              <div className='job-item__content'>
                <p className='job-item__title'>Report To</p>
                <p>
                  {reportTo?.name.split(' ')[0]}{' '}
                  {formatPhoneNumber(reportTo?.phone)}
                </p>
              </div>
            </div>
          )}
        </div>

        <div className='button-section'>
          {isLoading ? (
            <Spinner />
          ) : (
            <>
              <button
                onClick={handleRejectButton}
                className='button button-light'
              >
                No Thanks
              </button>
              <button
                onClick={handleAcceptButton}
                className='button button-dark'
              >
                "I'll Take it
              </button>
            </>
          )}
        </div>
      </div>
    </div>
  );
};

export default JobCard;
