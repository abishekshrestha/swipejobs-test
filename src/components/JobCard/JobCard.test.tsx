import { render } from '@testing-library/react';
import JobCard from './JobCard';

const jobData = {
  jobId: '5775d8e18a488e6c5bb08333',
  jobTitle: {
    name: 'Construction General Helper',
    imageUrl: 'https://imgs.swipejobs.com/js/job-category/construction-1.jpg',
  },
  company: {
    name: 'Steve Smith Construction',
    address: {
      formattedAddress: '430 Smith St, Chicago, IL 60654, USA',
      zoneId: 'America/Chicago',
    },
    reportTo: {
      name: 'Judy Smith',
      phone: '2130010012',
    },
  },
  wagePerHourInCents: 950,
  milesToTravel: 3.4,
  shifts: [
    {
      startDate: '2019-09-04T21:00:00Z',
      endDate: '2019-09-05T05:00:00Z',
    },
  ],
  branch: 'Downtown',
  branchPhoneNumber: '2531233322',
};

describe('JobCard', () => {
  it('should match snapshot', async () => {
    const { asFragment } = render(
      <JobCard
        {...jobData}
        handleAcceptButton={jest.fn()}
        handleRejectButton={jest.fn()}
        isLoading={false}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
