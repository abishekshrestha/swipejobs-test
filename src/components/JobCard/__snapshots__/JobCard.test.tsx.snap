// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobCard should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="job-card"
  >
    <img
      alt="job_image"
      class="image-style"
      src="https://imgs.swipejobs.com/js/job-category/construction-1.jpg"
    />
    <div
      class="job-description job-card-container"
    >
      <p
        class="job-description__title"
      >
        Construction General Helper
      </p>
      <p
        class="job-description__company"
      >
        Steve Smith Construction
      </p>
    </div>
    <div
      class="distance-rate-info job-card-container"
    >
      <div>
        <div
          class="distance-rate-info__title"
        >
          Distance
        </div>
        <div
          class="distance-rate-info__value"
        >
          3.4 miles
        </div>
      </div>
      <div>
        <div
          class="distance-rate-info__title text-align-end"
        >
          Hourly Rate
        </div>
        <div
          class="distance-rate-info__value text-align-end"
        >
          <sup
            class="distance-rate-info__dollar"
          >
            $
          </sup>
          9.50
        </div>
      </div>
    </div>
    <div
      class="job-card-container job-card-container--content"
    >
      <div>
        <div
          class="job-item"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium job-item__icon css-i4bv87-MuiSvgIcon-root"
            data-testid="CalendarMonthIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 4h-1V2h-2v2H8V2H6v2H5c-1.11 0-1.99.9-1.99 2L3 20c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 16H5V10h14v10zM9 14H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2zm-8 4H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2z"
            />
          </svg>
          <div
            class="job-item__content"
          >
            <p
              class="job-item__title"
            >
              Shift Dates
            </p>
            <div
              class="job-item__date"
            >
              <div>
                Sep 4, Wed 4:00 pm - 12:00 am CDT
              </div>
            </div>
          </div>
        </div>
        <a
          class="job-item job-item__location"
          href="https://map.google.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium job-item__icon css-i4bv87-MuiSvgIcon-root"
            data-testid="LocationOnIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
            />
          </svg>
          <div
            class="job-item__content"
          >
            <p
              class="job-item__title"
            >
              Location
            </p>
            <p
              class=""
            >
              430 Smith St, Chicago, IL 60654, USA
            </p>
            <p
              class="job-item__content--small"
            >
              3.40 miles from your job search location
            </p>
          </div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium job-item__chevron css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowForwardIosIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
            />
          </svg>
        </a>
        <div
          class="job-item"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium job-item__icon css-i4bv87-MuiSvgIcon-root"
            data-testid="AccountCircleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 4c1.93 0 3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5S10.07 6 12 6zm0 14c-2.03 0-4.43-.82-6.14-2.88C7.55 15.8 9.68 15 12 15s4.45.8 6.14 2.12C16.43 19.18 14.03 20 12 20z"
            />
          </svg>
          <div
            class="job-item__content"
          >
            <p
              class="job-item__title"
            >
              Report To
            </p>
            <p>
              Judy (213) 001 0012
            </p>
          </div>
        </div>
      </div>
      <div
        class="button-section"
      >
        <button
          class="button button-light"
        >
          No Thanks
        </button>
        <button
          class="button button-dark"
        >
          "I'll Take it
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
